utt/signature-processor/include/sigProcessor.hpp

```diff
@@ -0,0 +1,181 @@
+  }
+  ~SigProcessor();
+
+  /**
+   * @brief Initiates the protocol for computing the full signature for a given UTT object. This is a thread safe method
+   * and it can be called by any thread.
+   *
+   * @tparam T one of <Budget, Mint, Transaction>
+   * @param sigIds a vector of identifiers for the signature computing jobs
+   * @param data the UTT object which we want to sign of (one of <Budget, Mint, Transaction>)
+   * @param signer the UTT's coinsSigner object
+   * @param job_timeout a timeout for this job(s). If the timeout passes, the replica will broadcast the partial
+   * signature to all. The default is 1000 ms
+   */
+  template <typename T>
+  void processSignature(std::vector<uint64_t> sigIds,
+                        const T& data,
+                        std::shared_ptr<libutt::api::CoinsSigner> signer,
+                        uint64_t job_timeout = 1000) {
+    for (const auto& sigId : sigIds) {
+      jobs_.registerEntry(sigId, [signer, data](uint64_t rid, const libutt::api::types::Signature& sig) {
+        (void)sig;
+        (void)rid;
+        return true;
+      });
+    }
+    auto partial_sigs = signer->sign<T>(data);
+    handleSigJob(sigIds, partial_sigs, job_timeout);
+  }
+
+ private:
+  /**
+   * @brief Handles the event of receiving new partial signature
+   *
+   * @param sigId the sig ID
+   * @param sig_source the sig creator
+   * @param partial_sig the partial signature
+   */
+  void onReceivingNewPartialSig(uint64_t sigId, uint16_t sig_source, const libutt::api::types::Signature& partial_sig);
+
+  /**
+   * @brief Handles the event of receiving a new valid full signature. The validation assumed to be done by the caller.
+   *
+   * @param sigId the signature's job id
+   */
+  void onReceivingNewValidFullSig(uint64_t sigId);
+
+  /**
+   * @brief Handles the event of timeout for a specific job
+   *
+   * @param entry the entry that represents the job
+   */
+  void onJobTimeout(const SigJobEntry& entry);
+
+  /**
+   * @brief An internal helper method for computing signatures
+   *
+   * @param sigId the signature's job id
+   * @param partial_sig the partial signature
+   * @param job_timeout the timeout for this job
+   */
+  void handleSigJob(std::vector<uint64_t> sigId,
+                    const std::vector<libutt::api::types::Signature>& partial_sig,
+                    uint64_t job_timeout);
+
+  std::shared_ptr<bftEngine::impl::MsgsCommunicator> msgsCommunicator_;
+  concordUtil::Timers& timers_;
+  uint16_t repId_;
+  uint16_t n_;
+  uint16_t threshold_;
+  uint64_t timer_handler_timeout_;
+  JobCoordinatorCb getJobCoordinator = [&](uint64_t sigId) { return sigId % n_; };
+  GenerateAppClientRequestCb generateAppClientRequest = [](uint64_t sig_id, const libutt::api::types::Signature& sig) {
+    std::vector<uint8_t> ret(sizeof(uint64_t) + sig.size(), 0);
+    std::memcpy(ret.data(), &sig_id, sizeof(uint64_t));
+    std::memcpy(ret.data() + sizeof(uint64_t), sig.data(), sig.size());
```

Comment:
Probably need to consider endianness here for the multibyte values

Reply:
This will be probably used only for testing